File: lecture0-30/lecture012/subSeq.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;
vector<string> subSeq(string str)
{
    if (str.size() == 0)
    {
        vector<string> base;
        base.push_back(" ");
        return base;
    }

    char ch = str[0];
    string restStr = str.substr(1);

    vector<string> recAns = subSeq(restStr);
    vector<string> myAns;
    for (string s : recAns)
    {
        myAns.push_back(s);
        myAns.push_back(ch + s);
    }
    return myAns;
}
void solve()
{
    vector<string> ans = subSeq("1234");
    for (string s : ans)
        cout << s << " ";
}
int main(int args, char **argv)
{
    solve();
}

File: lecture012/subSeq.java
import java.util.Scanner;
import java.util.ArrayList;

public class subSeq {
    public static Scanner scn = new Scanner(System.in);

    public static void main(String[] args) {
        solve();
    }

    public static void solve() {
        System.out.println(subsequence("1234"));
    }

    public static ArrayList<String> subsequence(String str) {
        if (str.length() == 0) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }
        char ch = str.charAt(0);
        String restStr = str.substring(1);

        ArrayList<String> recAns = subsequence(restStr);
        ArrayList<String> myAns = new ArrayList<>();

        for (String s : recAns) {
            myAns.add(s);
            myAns.add(ch + s);
        }
        return myAns;
    }

}
